File: deal.py
from datetime import datetime

import utils

class Deal:
    """This class represents a daily deal and is initialized with
    data in JSON format

    """

    def __init__(self, data):
        self.id              = data['_id']
        self.item            = data['StoreItem']
        self.expiry          = datetime.fromtimestamp(data['Expiry']['sec'])
        self.original_price  = data['OriginalPrice']
        self.sale_price      = data['SalePrice']
        self.total           = data['AmountTotal']
        self.sold            = data['AmountSold']

    def __str__(self):
        """Returns a string with all the information about this alert

        """

        deal_string = ('Daily Deal: {0}\n'
                       '{1}p (original {2}p)\n'
                       '{3} / {4} sold\n'
                       'Expires in {5}')

        return deal_string.format(self.item, self.sale_price,
                                  self.original_price, self.sold,
                                  self.total, self.get_eta_string())

    def get_eta_string(self):
        """Returns a string containing the deal's ETA

        """
        return utils.timedelta_to_string(self.expiry - datetime.now())

File: src/lib/deal.js
var util = require('util');
var dsUtil = require('./_utils.js');
var lineEnd = process.env.HUBOT_LINE_END || '\n';
var blockEnd = process.env.HUBOT_BLOCK_END || ' ';

/**
 * Create a new deal instance
 *
 * @constructor
 * @param {object} data Deal data
 */
var Deal = function(data) {
  this.id = data._id;
  this.item = data.StoreItem;
  this.expiry = new Date(1000 * data.Expiry.sec);
  this.originalPrice = data.OriginalPrice;
  this.salePrice = data.SalePrice;
  this.total = data.AmountTotal;
  this.sold = data.AmountSold;
}

/** Return a string representation of the deal
 *
 * @return {string} The new string
 */
Deal.prototype.toString = function() {
  var dealString = util.format('Daily Deal: %s%s' +
                               '%dp (original %dp)%s' +
                               '%d / %d sold%s' +
                               'Expires in %s',
                               this.item, lineEnd,
                               this.salePrice, this.originalPrice, lineEnd,
                               this.sold, this.total, lineEnd,
                               this.getETAString(), blockEnd);
  return dealString;
}

/**
 * Return how much time is left before the deal expires
 *
 * @return {string} Deal ETA
 */
Deal.prototype.getETAString = function() {
  return dsUtil.timeDeltaToString(this.expiry.getTime() - Date.now());
}

module.exports = Deal;
